### DISCUSSION TITLE
How can you write exceptions without forcing Ignition to be installed?

### DISCUSSION THREAD

**Discussion by jasonvarga** [spatie/ignition#466] [Q&A]
In a package, how can you write an exception that implements `ProvidesSolutions` without requiring Ignition?

For example, in a Laravel package, I could have an exception like this:

```php
class MyException implements ProvidesSolutions
{
  //
}
```

Without Ignition installed, you get `Error: Interface "Spatie\Ignition\Contracts\ProvidesSolution" not found`.

We can fix that by requiring Ignition in the package, but that means we're forcing it onto people's projects.

In the past, we could just require the `ignition-contracts` package which prevents our error, and allows users to avoid installing Ignition itself if they don't want it. However, when you changed over the `spatie` namespace, you merged that package into this one. See https://github.com/spatie/ignition/issues/8

I obviously love Ignition and would recommend always using it, but maybe people had their reasons for avoiding it. They could simply remove it from their composer.json.

**Comment by jasonvarga** [comment]
@AlexVanderbist @freekmurze I'd be happy to contribute a PR to make this possible again, but of course there's no way for me to PR the creation of a new spatie package. 😅

**Comment by AlexVanderbist** [comment]
Hi, that's a good point! I wonder if a subtree split of the contracts namespace would be a good solution. Similar to how Laravel published their `Illuminte/*` packages as separate composer packages too: https://github.com/illuminate/http. That way we'd keep the single source code repo but have the benefits of a lightweight contract repo. 

I'll try to look into this later this week, as you said; it's quite impossible to PR a new repo :p

**Reply by jasonvarga** [reply]
Much appreciated!

**Comment by jasonvarga** [comment]
We realized that requiring `spatie/ignition` is okay since it doesn't automatically enable Ignition.

The `laravel-ignition` package would automatically enable Ignition in Laravel projects. Users are free to remove that if they want. Ignition itself would still technically be installed if we require it, but it'll basically just sit there disabled and do nothing.
